### PR TITLE
Add metrics

### PR DESCRIPTION
**Pull request by BobyMCbobs**
Add Prometheus client instrumentation for HTTP handler and new metrics port.

Metrics available:
- amount of requests
  - response code
  - method
  - version (v1, v2)
- amount of active requests

Metrics is exposed on a separate port as to not expose the info publicly.

### CODE REVIEW

**Comment by BenTheElder** [comment]
- How is the compatibility with cloud run? I don't recall METRICS_PORT.
- Do we even need custom code to obtain metrics like # of requests with cloud run?

https://cloud.google.com/run/docs/monitoring

File: cmd/archeio/main.go [hunk 1]
```diff
@@ -77,11 +138,14 @@ func handler(w http.ResponseWriter, r *http.Request) {
 	path := r.URL.Path
 	switch {
 	case strings.HasPrefix(path, "/v2/"):
+		serverMetrics.Requests.With(prometheus.Labels{"code": fmt.Sprintf("%v", http.StatusPermanentRedirect), "method": "GET", "version": "v2"}).Inc()
 		doV2(w, r)
 	case strings.HasPrefix(path, "/v1/"):
+		serverMetrics.Requests.With(prometheus.Labels{"code": fmt.Sprintf("%v", http.StatusPermanentRedirect), "method": "GET", "version": "v2"}).Inc()
```
Comment (aojea) [inline cmd/archeio/main.go:144]: is this always a GET?

Comment (BobyMCbobs) [reply cmd/archeio/main.go]: Whoops, that shouldn't be like that. Fixed in 3599385

**Comment by BobyMCbobs** [comment]
> * How is the compatibility with cloud run? I don't recall METRICS_PORT.
> * Do we even need custom code to obtain metrics like # of requests with cloud run?
> 
> https://cloud.google.com/run/docs/monitoring

_METRICS_PORT_ is an arbitrary environment variable. I'm still figuring out the integrations

File: cmd/archeio/main.go [hunk 2]
```diff
@@ -60,28 +103,50 @@ func main() {
...
 	// valid requests should be at /v2/ or /v1/, so we leave this check
 	// in the future we will selectively redirect clients to different copies
+	method := r.Method
 	path := r.URL.Path
 	switch {
 	case strings.HasPrefix(path, "/v2/"):
+		serverMetrics.Requests.With(prometheus.Labels{"code": fmt.Sprintf("%v", http.StatusPermanentRedirect), "method": method, "version": "v2"}).Inc()
 		doV2(w, r)
 	case strings.HasPrefix(path, "/v1/"):
+		serverMetrics.Requests.With(prometheus.Labels{"code": fmt.Sprintf("%v", http.StatusPermanentRedirect), "method": method, "version": "v2"}).Inc()
```
Comment (rikatz) [inline cmd/archeio/main.go:145]: Version=v1 in the end of the string?

Comment (BobyMCbobs) [reply cmd/archeio/main.go]: Thank you @rikatz, I missed this. Fixed in 3a50416

**Comment by BobyMCbobs** [comment]
> > * How is the compatibility with cloud run? I don't recall METRICS_PORT.
> > * Do we even need custom code to obtain metrics like # of requests with cloud run?
> > 
> > https://cloud.google.com/run/docs/monitoring
> 
> _METRICS_PORT_ is an arbitrary environment variable. I'm still figuring out the integrations

@BenTheElder, I've been trying to see how to integrate custom metrics through the instrumentation in this PR into Cloud Run.
The closest thing that I've been able to find is https://cloud.google.com/monitoring/custom-metrics/creating-metrics, however this doesn't mention Cloud Run.

May I please have some clarification for on what is being asked for, in https://github.com/kubernetes-sigs/oci-proxy/issues/17?
cc @BenTheElder @ameukam

**Comment by ameukam** [comment]
> May I please have some clarification for on what is being asked for, in #17? cc @BenTheElder @ameukam

@BobyMCbobs #17 was created firstly as a point of centralization of any discussion related to monitoring. This does not necessarily imply a technical action. Your PR introduces instrumentation using prometheus but the metrics introduced are already provided by the Cloud Run service like `BenTheElder`. IMHO, the PR can be closed and bring back the conversation when we are done with `registry.k8s.io` initial rollout.

**Comment by BobyMCbobs** [comment]
> > May I please have some clarification for on what is being asked for, in #17? cc @BenTheElder @ameukam
> 
> @BobyMCbobs #17 was created firstly as a point of centralization of any discussion related to monitoring. This does not necessarily imply a technical action. Your PR introduces instrumentation using prometheus but the metrics introduced are already provided by the Cloud Run service like `BenTheElder`. IMHO, the PR can be closed and bring back the conversation when we are done with `registry.k8s.io` initial rollout.

@ameukam, @dims asked me (see: https://kubernetes.slack.com/archives/CCK68P2Q2/p1647550187344329?thread_ts=1647548056.298549&cid=CCK68P2Q2). But I'll circle back after the initial rollout.

/close

**Pull request closed without merging**